File: 01-Programming-Languages/Java/Data-Structures/Tree/BinaryTreeTraversal.java
/*
 * Problem: Perform inorder, preorder, and postorder traversal of a binary tree.
 * Approach: Recursive traversal.
 * Time Complexity: O(n)
 * Space Complexity: O(n) (recursion stack)
 * Example: Tree: 1 -> left: 2, right: 3 -> Inorder: [2,1,3], Preorder: [1,2,3], Postorder: [2,3,1]
 */

import java.util.*;

public class BinaryTreeTraversal {
    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    public static List<Integer> inorder(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        inorderHelper(root, result);
        return result;
    }

    private static void inorderHelper(TreeNode node, List<Integer> result) {
        if (node == null) return;
        inorderHelper(node.left, result);
        result.add(node.val);
        inorderHelper(node.right, result);
    }

    public static List<Integer> preorder(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        preorderHelper(root, result);
        return result;
    }

    private static void preorderHelper(TreeNode node, List<Integer> result) {
        if (node == null) return;
        result.add(node.val);
        preorderHelper(node.left, result);
        preorderHelper(node.right, result);
    }

    public static List<Integer> postorder(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        postorderHelper(root, result);
        return result;
    }

    private static void postorderHelper(TreeNode node, List<Integer> result) {
        if (node == null) return;
        postorderHelper(node.left, result);
        postorderHelper(node.right, result);
        result.add(node.val);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        System.out.println("Inorder: " + inorder(root));
        System.out.println("Preorder: " + preorder(root));
        System.out.println("Postorder: " + postorder(root));
    }
}

File: 01-Programming-Languages/Kotlin/Data-Structures/Tree/binary_tree_traversal.kt
/*
 * Problem: Perform inorder, preorder, and postorder traversal of a binary tree.
 * Approach: Recursive traversal.
 * Time Complexity: O(n)
 * Space Complexity: O(n) (recursion stack)
 * Example: Tree: 1 -> left: 2, right: 3 -> Inorder: [2,1,3], Preorder: [1,2,3], Postorder: [2,3,1]
 */

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

fun inorder(root: TreeNode?): List<Int> {
    val result = mutableListOf<Int>()
    fun helper(node: TreeNode?) {
        if (node == null) return
        helper(node.left)
        result.add(node.`val`)
        helper(node.right)
    }
    helper(root)
    return result
}

fun preorder(root: TreeNode?): List<Int> {
    val result = mutableListOf<Int>()
    fun helper(node: TreeNode?) {
        if (node == null) return
        result.add(node.`val`)
        helper(node.left)
        helper(node.right)
    }
    helper(root)
    return result
}

fun postorder(root: TreeNode?): List<Int> {
    val result = mutableListOf<Int>()
    fun helper(node: TreeNode?) {
        if (node == null) return
        helper(node.left)
        helper(node.right)
        result.add(node.`val`)
    }
    helper(root)
    return result
}

fun main() {
    val root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    println("Inorder: ${inorder(root)}")
    println("Preorder: ${preorder(root)}")
    println("Postorder: ${postorder(root)}")
}
